### ISSUE TITLE
LDAP users missing email address 

### ISSUE THREAD

**Issue by childsc** [gocd/gocd#6619]
##### Issue Type

Bug

##### Summary

I see that for version 19.6 there could be a notification apli available. But we see that currently not all users email addresses are captured from our AD.  There is also no way for us to change this manually without deleting the user. 

I was unsure whether this issue should be added to GoCD or logged with the plugin.

This appears to be similar to the closed task 144 (this was closed due to inactivity) 

###### Basic environment details

* Go Version: 19.5.0
* JAVA Version: 11.0.1
* OS: Linux 3.13.0-147-generic
* Browser vendor and version (if relevant): Any

##### Steps to Reproduce

I am unsure how some email addresses are added when the user is created as all our users should be validated through ldap. (AD).

##### Expected Results
All users to have email address displayed.

##### Actual Results
a number of users have their email addresses missing either in the UI or when calling the get User api.

##### Possible Fix
Like with other services it would be nice to have a synch button. I believe that other services call it "Sxnc Directory"

##### Log snippets
As I am not able to add users with no email address I am not able to produce the logs.

**Comment by bdpiparva** [comment]
@childsc - Plugin uses the attribute configured as`EmailAttribute` to fetch the email from the LDAP server and will be mapped to the user in GoCD.

e.g You have to put `EmailAttribute=userMail` in auth config to map `neil@example.com` user in GoCD. Email mapping happens only once when the user is created in GoCD.

```
dn: uid=neil,ou=Employees,ou=Enterprise,ou=Principal,ou=system
objectClass: top
objectClass: person
objectClass: inetOrgPerson
objectClass: organizationalPerson
uid: neil
cn: Neil Parker
sn: Parker
userPassword: neil
displayName: Neil Parker
userMail: neil@example.com
l: New York
```

> There is also no way for us to change this manually without deleting the user.

You can update the user using [users](https://api.gocd.org/19.6.0/#users). 

```bash
curl 'https://ci.example.com/go/api/users/jdoe' \
      -u 'username:password' \
      -H 'Accept: application/vnd.go.cd.v3+json' \
      -H 'Content-Type: application/json' \
      -X PATCH \
      -d '{
            "email": "new-email@example.com",
            "email_me": true,
            "checkin_aliases": ["jdoe", "johndoe"]
          }'
```

The information provided in the issue description is not helping much. Please can you check/provide the following information(after reducting the sensitive info)

- Is password file plugin configured on the LDAP server?
  * If yes, make sure to use different username then the LDAP server to avoid confusion
- List the user's attributes from the LDAP server for the user -
  * having an email mapped in GoCD
  * an email is missing in GoCD 

See if both the user have email configured in the LDAP server and using the same email attributes.

**Comment by childsc** [comment]
hello @bdpiparva 

Thank you very much for your detailed response. 

I see in the configuration for the ldap plugin that EmailAttribute=userMail  should be the default?
`Value of the attribute is mapped to GoCD User's mail field, default value used is mail. `

![image](https://user-images.githubusercontent.com/40882907/61212813-93ec7b00-a703-11e9-90c9-103ec3c934bd.png)

I am not sure how long these low number of users with no email addresses have been there as this does not appear to be happening to the newer users. 

Here is the outputs from your request.
    Q. Is password file plugin configured on the LDAP server?  
        A. No this is configured on the local GoCD server.
    Q. List the user's attributes from the LDAP server for the user. 
        A. Please note I used the Powershell command "Get-ADUser <username> -Properties mail' (there is no property called userMail  in my response from the ldap server. 
        A. Email mapped:
```
DistinguishedName : CN=Smith\, John,OU=Externals,OU=Users,OU=Location,OU=DE,DC=grp,DC=Somegroup,DC=com
Enabled           : True
GivenName         : John
mail              : John.Smith@some-company.com
Name              : Smith, John
ObjectClass       : user
ObjectGUID        : c690da0a-96d7-4891-8d9f-54b06b5b6fb8
SamAccountName    : SmithJ
SID               : S-1-5-21-251489984-1497240538-811669034-19020
Surname           : Smith
UserPrincipalName : John.Smith@some-company.com

```
A. Email is missing in GoCD

```
DistinguishedName : CN=Blogs\, Joe,OU=Externals,OU=Users,OU=Location,OU=DE,DC=grp,DC=Somegroup,DC=com
Enabled           : True
GivenName         : Joe
mail              : Joe.Blogs@some-company.com
Name              : Blogs, Joe
ObjectClass       : user
ObjectGUID        : 0dfec06a-8a23-480e-ac6a-4efd1cb121b1
SamAccountName    : BlogsJ
SID               : S-1-5-21-251489984-1497240538-811669034-48171
Surname           : Blogs
UserPrincipalName : BlogsJ@some-company.com
```

I am assuming my best plan of action is to just update the users accounts manually.

**Comment by bdpiparva** [comment]
@childsc 

> I see in the configuration for the ldap plugin that EmailAttribute=userMail should be the default?

No, you don't have to change anything as in your LDAP server `mail` attribute is present. 

> I am assuming my best plan of action is to just update the users accounts manually.

Yes, you can update it manually if it is missing for the few users. It is reproducible on you LDAP server? It will be interesting to see why emails are missing for the few users?

**Comment by childsc** [comment]
Thanks @bdpiparva  

I will update my users and I will see if I can re-produce the issue tomorrow.

**Comment by childsc** [comment]
ok. after looking at the users, they all seem to be some of the first users. I can only assume they were added before the AD plugin was used.  

Thanks for your time. I will close the ticket.
